metaseq/benchmark/generator.py

```diff
@@ -0,0 +1,136 @@
+        print(f"Peak memory usage: {peak_memory / 1024 ** 3:.4f} GB")
+
+
+def benchmark(
+    batch_size: int, input_length: int, output_length: int
+) -> Tuple[np.ndarray, np.ndarray]:
+    time_elapsed = []
+    for _ in range(NUM_REPEATS):
+        start = torch.cuda.Event(enable_timing=True)
+        end = torch.cuda.Event(enable_timing=True)
+        start.record()
+
+        generator.generate(
+            inputs=np.random.randint(100, 50000, size=(batch_size, input_length)),
+            max_tokens=[output_length] * batch_size,
+            temperature=cfg.generation.temperature,
+            n=cfg.generation.beam,
+            top_p=0.9,
+        )
+        end.record()
+        torch.cuda.synchronize()
+        time_elapsed.append(start.elapsed_time(end))
+
+    peak_memory = torch.tensor(
+        torch.cuda.memory_stats()["allocated_bytes.all.peak"],
+        device=torch.cuda.current_device(),
+    )
+    torch.distributed.all_reduce(peak_memory)
+    return np.array(time_elapsed), peak_memory.cpu().numpy()
+
+
+def profile(batch_size: int, input_length: int, output_length: int) -> None:
+    tracing_schedule = schedule(skip_first=5, wait=5, warmup=5, active=2, repeat=1)
+    worker = socket.gethostname() + "_" + str(torch.distributed.get_rank())
+    trace_handler = tensorboard_trace_handler(
+        dir_name="traces", worker_name=worker, use_gzip=True
+    )
+
+    with torch.profiler.profile(
+        activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
+        profile_memory=True,
+        record_shapes=True,
+        schedule=tracing_schedule,
+        on_trace_ready=trace_handler,
+        with_stack=True,
+    ) as prof:
+        for _ in tqdm(range(17)):
+            generator.generate(
+                inputs=np.random.randint(100, 50000, size=(batch_size, input_length)),
+                max_tokens=[output_length] * batch_size,
+                temperature=cfg.generation.temperature,
+                n=cfg.generation.beam,
+                top_p=0.9,
+            )
+            prof.step()
+
+
+if __name__ == "__main__":
+    """
+    Example usage:
+        python metaseq/benchmark/generator.py \
+        --merges-filename /data/checkpoints/gpt2-merges.txt \
+        --vocab-filename /data/checkpoints/gpt2-vocab.json \
+        --path /data/checkpoints/opt-125m/reshard-no-os/reshard.pt \
+        --model-parallel-size 2 --distributed-world-size 2 \
+        --beam 1 --batch-size 4 --max-source-positions 4 --max-target-positions 16
+    """
+
+    parser = options.get_generation_parser()
+    parser.set_defaults(lr_scheduler=None, criterion=None)
+    LAUNCH_ARGS = ["--task language_modeling", "--bpe hf_byte_bpe", "/tmp"]
+    launch_args = sys.argv[1:] + [item for arg in LAUNCH_ARGS for item in arg.split()]
+    args = options.parse_args_and_arch(parser, input_args=launch_args)
+
+    args.bpe_merges = args.merges_filename
+    args.bpe_vocab = args.vocab_filename
+    cfg = convert_namespace_to_omegaconf(args)
```

Comment:
I added the following here:
```
model_overrides = {
        "memory_efficient_fp16": True,
        "fp16": True,
        "bf16": True,
        "sequence_parallel": False,
        "inference": True,
        "truncate_init": False
    }
cfg.common_eval.model_overrides = str(model_overrides)
```
Not sure if this is necessary. I can add this option as a follow up PR in case we want to load a bf16 checkpoint.